libraries/riscv-csr/src/csr.rs

```diff
@@ -212,4 +212,105 @@ impl<R: RegisterLongName, const V: usize> ReadWriteRiscvCsr<usize, R, V> {
     pub fn matches_all(&self, field: FieldValue<usize, R>) -> bool {
         field.matches_all(self.get())
     }
+
+    // Special methods only available on RISC-V CSRs, not found in the
+    // usual tock-registers interface
+
+    #[cfg(all(
+        any(target_arch = "riscv32", target_arch = "riscv64"),
+        target_os = "none"
+    ))]
+    #[inline]
+    pub fn atomic_replace(&self, val_to_set: usize) -> usize {
+        let r: usize;
+        unsafe {
+            asm!("csrrw {rd}, {csr}, {rs1}",
```

Comment:
I am not super familiar with the RISC-V ISA, but it is somewhat surprising to me that this compiles on earlgray despite it not having the A extension. Looking at the spec, I guess it is the case that the base instruction set includes atomic instructions for modifying CSRs, but not for general atomic operations?

Reply:
I think this is because CSRs are core-local, so atomicity is much easier to enforce here, vs. on memory in general. But yes, as far as I understand every compliant rv32i core has these instructions, and the "regular" CSR reads & writes are just pseudo instructions for these atomic ones.

Reply:
I also tested the csrrw, csrrs and csrrc on the VexRiscv without any extensions (just rv32i), but not this PR in its entirety yet, will do in a few hours.